File: lib/scraper/src/api/snapshot.ts
import { mkdir, readFile, rename, writeFile } from 'node:fs/promises';
import { dirname, resolve } from 'node:path';
import { fetchEntitySnapshot, type EntitySnapshot } from './deadlock';
import { entitySnapshotSchema } from '../types/deadlockApi';

export function entitySnapshotPath(): string {
	return resolve(process.env.CHANGELOGS_DIR || 'app/changelogs', 'entities.json');
}

const SNAPSHOT_SCHEMA_VERSION = 2;

export function serializeEntitySnapshot(snapshot: EntitySnapshot): string {
	const { heroes, items, provenance } = snapshot;
	const envelope = provenance
		? { schemaVersion: SNAPSHOT_SCHEMA_VERSION, ...provenance, heroes, items }
		: { heroes, items };
	return `${JSON.stringify(envelope, null, '\t')}\n`;
}

export async function writeEntitySnapshot(snapshot: EntitySnapshot): Promise<string> {
	const target = entitySnapshotPath();
	const staging = `${target}.${process.pid}.tmp`;
	await mkdir(dirname(target), { recursive: true });
	await writeFile(staging, serializeEntitySnapshot(snapshot));
	await rename(staging, target);
	return target;
}

export async function readEntitySnapshot(): Promise<EntitySnapshot | null> {
	let raw: string;
	try {
		raw = await readFile(entitySnapshotPath(), 'utf-8');
	} catch {
		return null;
	}
	let json: unknown;
	try {
		json = JSON.parse(raw);
	} catch {
		json = null;
	}
	const parsed = entitySnapshotSchema.safeParse(json);
	if (!parsed.success) {
		console.warn(`   ⚠️  ${entitySnapshotPath()} is not a usable snapshot`);
		return null;
	}
	const { heroes, items, schemaVersion, clientVersion, language, collectedAt } =
		parsed.data;
	if (schemaVersion === undefined || !clientVersion || !language || !collectedAt) {
		return { heroes, items };
	}
	return { heroes, items, provenance: { clientVersion, language, collectedAt } };
}

const withoutCollectionTime = ({ heroes, items, provenance }: EntitySnapshot): string =>
	JSON.stringify({
		clientVersion: provenance?.clientVersion,
		language: provenance?.language,
		heroes,
		items
	});

export async function loadEntitySnapshot(): Promise<EntitySnapshot> {
	let cached: EntitySnapshot | null | undefined;
	try {
		const snapshot = await fetchEntitySnapshot();
		cached = await readEntitySnapshot();
		if (cached && withoutCollectionTime(cached) === withoutCollectionTime(snapshot)) {
			return cached;
		}
		await writeEntitySnapshot(snapshot);
		return snapshot;
	} catch (error) {
		cached ??= await readEntitySnapshot();
		if (!cached) throw error;
		const reason = error instanceof Error ? error.message : String(error);
		console.warn(`   ⚠️  Deadlock asset API unreachable (${reason})`);
		console.warn(`   ↩️  Falling back to ${entitySnapshotPath()}`);
		return cached;
	}
}

File: crates/scraper/src/api/snapshot.rs
use std::path::{Path, PathBuf};

use anyhow::Result;
use deadlog_db::write::snapshot::{EntitySnapshot, read_snapshot, serialize_snapshot, without_collection_time};

use super::deadlock::fetch_entity_snapshot;
use crate::http::Http;

pub fn entity_snapshot_path(changelogs_dir: &Path) -> PathBuf {
    changelogs_dir.join("entities.json")
}

/// Writes through a staging file so a crash cannot leave a half-written snapshot.
pub fn write_entity_snapshot(changelogs_dir: &Path, snapshot: &EntitySnapshot) -> Result<PathBuf> {
    let target = entity_snapshot_path(changelogs_dir);
    let staging = PathBuf::from(format!("{}.{}.tmp", target.display(), std::process::id()));
    std::fs::create_dir_all(changelogs_dir)?;
    std::fs::write(&staging, serialize_snapshot(snapshot)?)?;
    std::fs::rename(&staging, &target)?;
    Ok(target)
}

/// The committed snapshot, or `None` when it is missing or not usable.
pub fn read_entity_snapshot(changelogs_dir: &Path) -> Option<EntitySnapshot> {
    let path = entity_snapshot_path(changelogs_dir);
    let raw = std::fs::read_to_string(&path).ok()?;
    match read_snapshot(&raw) {
        Ok(snapshot) => Some(snapshot),
        Err(_) => {
            eprintln!("   ⚠️  {} is not a usable snapshot", path.display());
            None
        }
    }
}

/// Heroes and items for this build: the live API when it answers, refreshing
/// `entities.json` when the API moved on; otherwise the committed snapshot, so an API
/// outage never blocks a build. A refresh that differs only in its collection time
/// keeps the committed file, and its timestamp, untouched.
pub fn load_entity_snapshot(http: &dyn Http, changelogs_dir: &Path, now: &str) -> Result<EntitySnapshot> {
    let fetched = fetch_entity_snapshot(http, now);
    let cached = read_entity_snapshot(changelogs_dir);
    match fetched {
        Ok(snapshot) => {
            if let Some(cached) = cached
                && without_collection_time(&cached)? == without_collection_time(&snapshot)?
            {
                return Ok(cached);
            }
            write_entity_snapshot(changelogs_dir, &snapshot)?;
            Ok(snapshot)
        }
        Err(error) => {
            let Some(cached) = cached else { return Err(error) };
            eprintln!("   ⚠️  Deadlock asset API unreachable ({error})");
            eprintln!("   ↩️  Falling back to {}", entity_snapshot_path(changelogs_dir).display());
            Ok(cached)
        }
    }
}
